File: Sliding Window/trappingRain.cpp
/* 
The amount of rain trapped above each index is equal to:
    min(max height of a peak to left of index, max height of a peak to right of index) - height at index
We could use a brute force approach to find each of these variables: 
    for every index we send a pointer out to the left looking for height of max peak to the left
    and vice versa for right

This would take O(n^2) time complexity though. We want to do it in O(n) and in one pass

So take a shorcut
Rather asking question "What is the max height of a peak to right of this index?"
We instead ask "Is there a peak to the right of this index that's at least as big as our current leftMax?"
If the answer is "Yes", then we actually don't care if there is a rightMax that is bigger than that
So long as it's at least as big as our current leftMax we can guarantee that there is a region where some water can be tapped

We use two pointers 
one starting from the left and and one starting from the right
both travelling inward
to keep track of leftMax and rightMax respectively

And as long as the max value a pointer has seen is less than the max value of it's opposite
then that pointer can proceed inwards

Time complexity         O(n)
Space complexity        O(1)
 */

class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        
        if (n < 3) return 0;
        

        int left(0), right(n - 1), leftMax(0), rightMax(0), res(0);
        
        while (left < right){
            leftMax = max(leftMax, height[left]);
            rightMax = max(rightMax, height[right]);
            if (leftMax <= rightMax){
                res += leftMax - height[left];
                ++left;
            }
            else{
                res += rightMax - height[right];
                --right;
            }
        }
        
        return res;
    }
};

File: Sliding Window/trappingRain.py
# https://leetcode.com/problems/trapping-rain-water/

# Time complexity         O(n)
# Space complexity        O(1)

class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        
        # Smallest region where rain can be trapped is 101
        if n < 3:
            return 0
        
        # Rain can only be trapped between two consecutive local maxmima
        # where the the height of one is equal to or greater than the height of the other

        # With the sliding window approach we use below
        # unless every other local maxima is the same height as the global maximum
        # then by defintion all other local maxima will not be equal to or greater height 

        # To resolve this, we split the input array in half at the global maximum
        # Sweep left to right in the left half
        # And right to left in the right half
        maxPeak = max(height)
        idxOfMax = height.index(maxPeak)
        
        l2r = 0
        front = 0
        lowerEdge = 0
        for back in range(idxOfMax + 1):
            # Front pointer advances looking for a peak
            while front < idxOfMax and (height[front] < height[back] or front == back):
                front += 1
                # When it finds one, it stores the height of the rear of the two consecutive peaks in lowerEdge
                lowerEdge = height[back]
            if height[back] < lowerEdge:
                # As the back pointer advances, it tallies the depth of rain trapped between the two regions
                l2r += lowerEdge - height[back]
                   
        r2l = 0
        back = n - 1
        for front in range(n - 1, idxOfMax - 1,  - 1):
            while back > idxOfMax and (height[front] > height[back] or front == back):
                back -= 1
                lowerEdge = height[front]
            if height[front] < lowerEdge:
                r2l += lowerEdge - height[front]
                
        return l2r + r2l